### PR TITLE
json: remove use of StatementContext, simplify mapper / argument lambdas

### PR DESCRIPTION
**Pull request by stevenschlansker**
we actually only need the ConfigRegistry instead, and that lets us move
a fair amount of work out of the lambdas in the json handling, which should be a minor performance improvement

### CODE REVIEW

**Comment by TheRealMarnes** [comment]
I'm not really seeing it, could you explain the issue, and which change lies at its source?

**Comment by stevenschlansker** [comment]
It's just a simplification.  Moved a lot of code out of the lambdas and do the work ahead of time.  This way, we don't e.g. re-lookup the mapper every time we call the lambda.  It also removes our need to get the StatementContext god-object and instead we can just use the ConfigRegistry.

File: json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java [hunk 1]
```diff
@@ -40,17 +41,12 @@
         if (String.class.equals(type)) {
             return Optional.empty();
         }
-
-        return Optional.of((pos, stmt, ctx) -> {
```
Comment (TheRealMarnes) [inline json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: I believe this change makes the json mapper rely on the String mapper present at the time of registering the json mapper, and ignore any string mapper registered afterward. Not a huge problem I think but maybe users need to be aware registration order will matter between those two?

File: json/src/main/java/org/jdbi/v3/json/internal/JsonColumnMapperFactory.java [hunk 2]
```diff
@@ -40,17 +42,17 @@
...
+        // look for specialized json support first, revert to simple String mapping if absent
+        ColumnMapper<String> jsonStringMapper = JdbiOptionals.findFirstPresent(
+                () -> cm.findFor(QualifiedType.of(String.class).with(Json.class)),
+                () -> cm.findFor(String.class))
+                .orElseThrow(() -> new UnableToProduceResultException(JSON_NOT_RETRIEVABLE));
 
+        final JsonMapper mapper = config.get(JsonConfig.class).getJsonMapper();
         return Optional.of((rs, i, ctx) -> {
-            // look for specialized json support first, revert to simple String mapping if absent
-            ColumnMapper<String> jsonStringMapper = JdbiOptionals.findFirstPresent(
```
Comment (TheRealMarnes) [inline json/src/main/java/org/jdbi/v3/json/internal/JsonColumnMapperFactory.java]: idem

File: json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java (hunk 1 above)
Comment (stevenschlansker) [reply json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: `ArgumentFactory` is only invoked once we are preparing the statement to execute - you have the (static) argument type at this point, this is far after registration time.

Comment (TheRealMarnes) [reply json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: Right, I'm stupid, sorry

Comment (qualidafial) [reply json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: I think it's reasonable to expect the `@Json String` argument factory and `JsonMapper` to be configured prior to executing a statement that depends on it.

Comment (qualidafial) [reply json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: @stevenschlansker That's not accurate. The `ArgumentFactory.build()` method is invoked by `Arguments.findFor()` which is invoked at the specific chaining method where you bind the value, e.g.

```
handle.createUpdate(sql)
    .bindByType("foo", foo, QualifiedType.of(Foo.class).with(Json.class)) // argument factory invoked here
    .execute();
```

So it's conceivable the if somebody configured the `JsonMapper` or registered the `@Json String` mapper later in the call chain, they wouldn't be used:

```
handle.createUpdate(sql)
    .bindByType("foo", foo, QualifiedType.of(Foo.class).with(Json.class))
    .configure(JsonConfig.class, jc -> jc.setJsonMapper(myMapper)) // too late!
    .execute();
```

I just don't think it's a very likely case, and not worth worrying about unless we hear a bug report on it.

I suspect the vast majority of folks using JSON mapping will just follow the cookbook in the developer guide and configure everything on the `Jdbi` instance.

Comment (stevenschlansker) [reply json/src/main/java/org/jdbi/v3/json/internal/JsonArgumentFactory.java]: Ah, sorry, I misspoke.  Either way my overall point was that you are already logically building the statement, even if we are not actually building the `PreparedStatement`.

**Pull request merged**